src/actions/channels.test.js

```diff
@@ -2125,6 +2131,49 @@ describe('Actions.Channels', () => {
             expect(state.entities.channels.myMembers[channel.id]).not.toBeDefined();
             expect(state.entities.channelCategories.byId[category.id].channel_ids).toEqual([]);
         });
+
+        test('should restore a channel when failing to leave it (non-custom category)', async () => {
+            const channel = {id: 'channel', team_id: team.id, type: General.OPEN_CHANNEL};
+            const category = {id: 'category', team_id: team.id, type: CategoryTypes.CHANNELS, channel_ids: [channel.id]};
+
+            store = await configureStore({
+                entities: {
+                    channelCategories: {
+                        byId: {
+                            category,
+                        },
+                        orderByTeam: {
+                            [team.id]: [category.id],
+                        },
+                    },
+                    channels: {
+                        channels: {
+                            channel,
+                        },
+                        myMembers: {
+                            [channel.id]: {channel_id: channel.id, user_id: user.id},
+                        },
+                    },
+                    users: {
+                        currentUserId: user.id,
+                    },
+                },
+            });
+
+            nock(Client4.getBaseRoute()).
+                delete(`/channels/${channel.id}/members/${user.id}`).
+                reply(500, {});
+
+            await store.dispatch(Actions.leaveChannel(channel.id));
+
+            const state = store.getState();
+
+            setTimeout(() => {
```

Comment:
Actually, you could probably get away with `await Promise.resolve(true)` to further simplify it since we only need to preempt this code to let the async block in the action run.

Reply:
Ooh, I like the promise approach instead of wrapping. However, I tried the `await`, and it failed for one test case, I think because of the interdependence on mocking the responses which probably requires a chain of pre-emptions.